feat: get variation details by variation type

duyhungtnn opened the pull request:
Related https://github.com/bucketeer-io/bucketeer/issues/886
### Changes
- Added a new data class `BKTEvaluationDetail<T>` to support getting variation details by variation type.
- Added a new data class `BKTValue` to reference [a JSON node value ](https://github.com/bucketeer-io/ios-client-sdk/pull/90/files#diff-a3a4eddf801ddca832aa6a192ebfb577b04567c5e6e08efa273d48be854ea950R6)
- Introduce a new interface to get variation details by variation type
```swift
func objectVariation(featureId: String, defaultValue: BKTValue) -> BKTValue
func intVariationDetails(featureId: String, defaultValue: Int) -> BKTEvaluationDetail<Int> {}
func doubleVariationDetails(featureId: String, defaultValue: Double) -> BKTEvaluationDetail<Double> {}
func boolVariationDetails(featureId: String, defaultValue: Bool) -> BKTEvaluationDetail<Bool> {}
func stringVariationDetails(featureId: String, defaultValue: String) -> BKTEvaluationDetail<String> {}
func objectVariationDetails(featureId: String, defaultValue: BKTValue) -> BKTEvaluationDetail<BKTValue> {}
```
- The class `BKTEvaluation` & the methods `getEvaluationDetails`, `jsonVariation` will be deprecated after this PR.

duyhungtnn committed:
feat: new detailed evaluation interface

duyhungtnn committed:
feat: add type generic method to new interface

duyhungtnn committed:
fix: swiftlint

duyhungtnn committed:
test: init new test BKTClientEvaluationDetailTests

duyhungtnn committed:
test: full unit tests for new interface

duyhungtnn committed:
fix: test fail

duyhungtnn committed:
fix: test json evaluation detail fail

duyhungtnn committed:
fix: remove incorrect type check

duyhungtnn committed:
fix: update deprecated message

duyhungtnn commented:
@cre8ivejp please help me to take a look on this PR

duyhungtnn committed:
chore: correct grammar

kakcy commented on Bucketeer/Sources/Public/BKTClient.swift:
```diff
@@ -12,23 +12,34 @@ public class BKTClient {
+    }
+
+    func getVariationDetail<T:Equatable>(featureId: String, defaultValue: T) -> BKTEvaluationDetails<T> {
         component.config.logger?.debug(message: "BKTClient.getVariation(featureId = \(featureId), defaultValue = \(defaultValue) called")
```
```suggestion
        component.config.logger?.debug(message: "BKTClient.getVariationDetail(featureId = \(featureId), defaultValue = \(defaultValue) called")
```

kakcy commented on Bucketeer/Sources/Public/BKTEvaluation.swift:
```diff
@@ -19,3 +21,64 @@ public struct BKTEvaluation: Equatable {
+    public let variationValue: T
+    public let reason: Reason
+
+    public enum Reason: String, Codable, Hashable {
```
Although it is similar to the [ReasonType](https://github.com/bucketeer-io/ios-client-sdk/blob/main/Bucketeer/Sources/Internal/Model/ReasonType.swift#L3C6-L3C16) definition, it is not very desirable to have the same definition in multiple places, so I would like you to consider using it uniformly if possible. Or consider removing the definition if you don't need it.

duyhungtnn replied:
[ReasonType](https://github.com/bucketeer-io/ios-client-sdk/blob/main/Bucketeer/Sources/Internal/Model/ReasonType.swift#L3C6-L3C16)

is for internal use only. I maintain the same approach like before

duyhungtnn replied:
I would like to keep this. The Android SDK has something like this as well. @kakcy

kakcy commented on Bucketeer/Sources/Public/BKTEvaluation.swift:
```diff
@@ -19,3 +21,64 @@ public struct BKTEvaluation: Equatable {
+        case offVariation = "OFF_VARIATION"
+        case prerequisite = "PREREQUISITE"
+
+        public static func fromString(value: String) -> Reason {
```
It might be good to have a test for the `fromString()`.

duyhungtnn replied:
I will add them

duyhungtnn replied:
updated in https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/d170ebe84a91bcc0b47b31cbfe310074d5e7cadd

kakcy commented on Bucketeer/Sources/Public/BKTEvaluation.swift:
```diff
@@ -1,5 +1,6 @@
 import Foundation
 
+@available(*, deprecated, message: "Use BKTEvaluationDetail<T> instead")
```
```suggestion
@available(*, deprecated, message: "Use BKTEvaluationDetails<T> instead")
```

kakcy commented on Bucketeer/Sources/Public/BKTEvaluation.swift:
```diff
@@ -10,6 +11,7 @@ public struct BKTEvaluation: Equatable {
     public let variationValue: String
     public let reason: Reason
 
+    @available(*, deprecated, message: "Use BKTEvaluationDetail<T>.Reason instead")
```
```suggestion
    @available(*, deprecated, message: "Use BKTEvaluationDetails<T>.Reason instead")
```

kakcy commented on Bucketeer/Sources/Public/BKTClient.swift:
```diff
@@ -200,6 +238,7 @@ extension BKTClient {
         }
     }
 
+    @available(*, deprecated, message: "use stringEvaluationDetails<T> instead")
```
??
```suggestion
    @available(*, deprecated, message: "use getVariationDetail<T> instead")
```

duyhungtnn replied:
hi @kakcy , 

`getVariationDetail` is for internal use only. It can not be accessed from outside.
But I will remote the `<T>` from the `stringEvaluationDetails<T>`

duyhungtnn replied:
I updated it to 
```swift
@available(*, deprecated, message: "use stringVariationDetails(featureId:, defaultValue:) instead")
```

kakcy reviewed (commented):
thank you!
I have made some comments, please check them out.

duyhungtnn committed:
Update Bucketeer/Sources/Public/BKTEvaluation.swift

Co-authored-by: kakcy <43775118+kakcy@users.noreply.github.com>

duyhungtnn committed:
Update Bucketeer/Sources/Public/BKTEvaluation.swift

Co-authored-by: kakcy <43775118+kakcy@users.noreply.github.com>

duyhungtnn committed:
Update Bucketeer/Sources/Public/BKTClient.swift

Co-authored-by: kakcy <43775118+kakcy@users.noreply.github.com>

duyhungtnn committed:
chore: correct deprecated message

duyhungtnn committed:
test: for Reason.fromString

duyhungtnn committed:
chore: update new interface method names

duyhungtnn committed:
chore: correct message text

duyhungtnn commented:
@kakcy thank you for reviewing.
I updated the PR with some changes from your suggestion.
Also, I changed the interface a little in this commit https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/12cbd20de5e303e36e8c4fce890ea8cbaf5fce39

as they are a new requirement https://github.com/bucketeer-io/android-client-sdk/pull/180#issuecomment-2188632955

kakcy reviewed (dismissed):
Thank you for your responses 👍

cre8ivejp reviewed (dismissed):
Thank you!
I'll merge next week.

duyhungtnn committed:
feat: support JSON Array

duyhungtnn committed:
chore: update comments

duyhungtnn committed:
test: add BKTValue tests

duyhungtnn commented:
hi @cre8ivejp , @kakcy 
Please help me to take a look at this PR. 
I added changes from the latest review. These changes make our interface closest to OpenFeatue specification 

# Changes 
- Added a new data class `BKTValue` to reference [a JSON node value](https://github.com/bucketeer-io/ios-client-sdk/pull/90/files#diff-a3a4eddf801ddca832aa6a192ebfb577b04567c5e6e08efa273d48be854ea950R6)
- Added [a new method](https://github.com/bucketeer-io/ios-client-sdk/pull/90/files#diff-2edc41fe19545933e618fc1376f5028e229b74f3e7c301ab02251976e3f915d3R162) to return a `BKTValue`
```
public func objectVariationDetails(featureId: String, defaultValue:BKTValue)
```
- Removed `jsonVariationDetail`
- Deprecated the method `jsonVariation`

duyhungtnn committed:
chore: make logic to decode BKTValue more simpler

duyhungtnn committed:
chore: add test testGetVariationValueAsBKTValue

duyhungtnn commented:
Although this PR seems ready for review. However, I found some edge cases when finalizing the same changes in the Android PR.

I will add some new tests to this PR.

duyhungtnn committed:
tests: update BKTValueTests

duyhungtnn committed:
feat: Add support to return intValue even when the underlying value is a double

duyhungtnn commented:
### New changes 
- Added support to return intValue even when the underlying value is a double https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/8d576f2c755d14578a826afcea8d8af9d42f3fd4
- Created BKTValue.number to hold JSON number value https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/58a6bb7a99e5792fcc17e2978b90ba4ac8d77baa

duyhungtnn committed:
feat: create BKTValue.number to hold JSON number value

duyhungtnn committed:
tests: add more test for BKTValueTests.swift

duyhungtnn committed:
chore: keep the detail interface right after the non-detail interface

duyhungtnn commented:
### New changes
- keep the detail interface right after the non-detail interface https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/3515534d3a7faae2f048afce54bad82c31c9dc76

cre8ivejp commented on Bucketeer/Sources/Public/BKTClient.swift:
```diff
@@ -126,24 +143,50 @@ extension BKTClient {
+
+    public func stringVariationDetails(featureId: String, defaultValue: String) -> BKTEvaluationDetails<String> {
+        return getVariationDetail(featureId: featureId, defaultValue: defaultValue)
+    }
```
Can we call the same function for both cases?
It's a bit confusing. The non-details functions call the `stringVariationDetails` but the details function calls a different function `getVariationDetail`.

E.g.
```suggestion
    public func stringVariation(featureId: String, defaultValue: String) -> String {
        return stringVariationDetails(featureId: featureId, defaultValue: defaultValue).variationValue
    }

    public func stringVariationDetails(featureId: String, defaultValue: String) -> BKTEvaluationDetails<String> {
        return stringVariationDetails(featureId: featureId, defaultValue: defaultValue)
    }
```

cre8ivejp replied:
Like this.
https://github.com/bucketeer-io/go-server-sdk/pull/136/files#diff-9658208f158246a9cfbc2c798e77e8224eaed053c42d140eb1e7237d8ed4e91cR295-R305

duyhungtnn committed:
chore: rename `getVariationDetail` to `getBKTEvaluationDetails`

duyhungtnn committed:
Merge branch 'main' into feat/new-detailed-evaluation-interface

duyhungtnn committed:
fix: build fail

cre8ivejp commented on Bucketeer/Sources/Public/BKTEvaluation.swift:
```diff
@@ -19,3 +21,64 @@ public struct BKTEvaluation: Equatable {
+    public let variationValue: T
+    public let reason: Reason
+
+    public enum Reason: String, Codable, Hashable {
```
@duyhungtnn, WDYT of adding the `BKTEvaluationDetails` to a new file like you did in the Android SDK?

duyhungtnn replied:
It's good.
I updated https://github.com/bucketeer-io/ios-client-sdk/pull/90/commits/6e761aca5f0edf4453bcf9e5778f0d7d418916b8

duyhungtnn committed:
chore: move BKTEvaluationDetails to a new file

cre8ivejp reviewed (approved):
Thank you!
